File: main/rotate-cities/rotate_cities.rb
#!/usr/bin/env ruby
$VERBOSE = true

def each_rotation(text)
  text.length.times { |i| yield text[i..-1] + text[0...i] }
end

def find_or_create_row(results, map, key)
  row = map[key]
  return row unless row.nil?

  row = []
  results << row
  each_rotation(key) { |rot| map[rot] = row }

  row
end

def group_by_rotations(cities)
  results = []
  map = {}

  cities.each do |city|
    find_or_create_row(results, map, city.downcase.freeze) << city
  end

  results
end

a = %w[Tokyo London Rome Donlon Kyoto Paris]
p group_by_rotations(a)

File: main/rotate-cities/rotate-cities.py
#!/usr/bin/env python3

def rotations(text):
    for i in range(len(text)):
        yield text[i:] + text[:i]

def partition_by_rotations(cities):
    results = []
    lookup = {}

    for city in cities:
        key = city.casefold()

        try:
            row = lookup[key]
        except KeyError:
            row = []
            results.append(row)
            for rot in rotations(key):
                lookup[rot] = row

        row.append(city)

    return results

a = ['Tokyo', 'London', 'Rome', 'Donlon', 'Kyoto', 'Paris']
print(partition_by_rotations(a))
